src/plots/VolumeVTK9/avtVolumeFilter.C

```diff
@@ -0,0 +1,1462 @@
+                               (bounds[3]-bounds[2])*(bounds[3]-bounds[2]) +
+                               (bounds[5]-bounds[4])*(bounds[5]-bounds[4]));
+            range[1] = (actualRange[1]*diag) / 2.;
+    */
+            range[1] = (actualRange[1]) / 4.;
+        }
+    }
+
+
+    return om;
+}
+
+// ****************************************************************************
+//  Method: GetLogicalBounds
+//
+//  Purpose:
+//      Added for no resampling for VTK_RECTILINEAR_GRID data of more
+//      than 1 leaf.
+//
+// ****************************************************************************
+
+bool
+avtVolumeFilter::GetLogicalBounds(avtDataObject_p input,
+                                  int &width,int &height, int &depth)
+{
+    const avtDataAttributes &datts = input->GetInfo().GetAttributes();
+    std::string db = input->GetInfo().GetAttributes().GetFullDBName();
+
+    debug5<<"datts->GetTime(): "<<datts.GetTime()<<endl;
+    debug5<<"datts->GetTimeIndex(): "<<datts.GetTimeIndex()<<endl;
+    debug5<<"datts->GetCycle(): "<<datts.GetCycle()<<endl;
+
+    ref_ptr<avtDatabase> dbp = avtCallback::GetDatabase(db, datts.GetTimeIndex(), nullptr);
+    avtDatabaseMetaData *md = dbp->GetMetaData(datts.GetTimeIndex(), 1);
+    std::string mesh = md->MeshForVar(datts.GetVariableName());
+    const avtMeshMetaData *mmd = md->GetMesh(mesh);
+
+    if (mmd->hasLogicalBounds == true)
+    {
+        width=mmd->logicalBounds[0];
+        height=mmd->logicalBounds[1];
+        depth=mmd->logicalBounds[2];
+
+        return true;
+    }
+
+    return false;
+}
+
+// ****************************************************************************
+//  Method: avtVolumeFilter::GetRenderVariables
+//
+//  Purpose:
+//      Determine which variables to use for the rendering.
+//
+//  Programmer:
+//  Creation:
+//
+// ****************************************************************************
+int
+avtVolumeFilter::GetRenderVariables( int &primIndex,
+                                     int &opacIndex,
+                                     int &gradIndex,
+                                     char *gradName)
+{
+    // Determine which variables to use for the rendering.
+    avtDataset_p input = GetTypedInput();
+
+    primIndex = -1;
+    opacIndex = -1;
+    gradIndex = -1;
+
+    int count = 0;
+
+    const char *gradvar = atts.GetOpacityVariable().c_str();
+    if (strcmp(gradvar, "default") == 0)
+        gradvar = primaryVariable;
+
+    VarList vl;
+    avtDatasetExaminer::GetVariableList(input, vl);
+
+    // This name is explicitly sent to the avtGradientExpression in
+    // avtVolumePlot.
+    snprintf(gradName, 128, "_%s_gradient", gradvar);
+
+    for (int i = 0 ; i < vl.nvars ; i++)
+    {
+        if ((strstr(vl.varnames[i].c_str(), "vtk") != nullptr) &&
+            (strstr(vl.varnames[i].c_str(), "avt") != nullptr))
+            continue;
+
+        if (vl.varnames[i] == primaryVariable)
+        {
```

Comment:
Just an FYI (because it relates to prev. comment but has nothing to do with proposing you change anything here)...I also try to avoid a scoped `if` with a single line body. But, thats just me.